esignet-core/src/main/java/io/mosip/esignet/core/dto/ProofKeyCodeExchange.java

```diff
@@ -0,0 +1,61 @@
+package io.mosip.esignet.core.dto;
+
+import io.mosip.esignet.core.constants.ErrorConstants;
+import io.mosip.esignet.core.exception.EsignetException;
+import io.mosip.esignet.core.util.IdentityProviderUtil;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.util.StringUtils;
+
+import java.nio.charset.Charset;
+
+/**
+ * PKCE support for OIDC transaction when using authorization code flow.
+ * technique to mitigate auth-code interception attacks.
+ */
+@Slf4j
+@Getter
+@Setter
+public class ProofKeyCodeExchange {
+
+    private static final String S256 = "S256";
+
+    private String codeChallenge;
+    private String codeChallengeMethod;
+
+    private ProofKeyCodeExchange() {}
+
+    public static ProofKeyCodeExchange getInstance(String codeChallenge, String codeChallengeMethod) {
+        if(codeChallengeMethod == null || codeChallenge == null)
+            return null;
+
+        if(StringUtils.isEmpty(codeChallenge))
+            throw new EsignetException(ErrorConstants.INVALID_PKCE_CHALLENGE);
+
+        switch (codeChallengeMethod) {
+            case S256 :
+                ProofKeyCodeExchange proofKeyCodeExchange = new ProofKeyCodeExchange();
+                proofKeyCodeExchange.setCodeChallenge(codeChallenge);
+                proofKeyCodeExchange.setCodeChallengeMethod("S256");
+                return proofKeyCodeExchange;
+            default:
+                throw new EsignetException(ErrorConstants.UNSUPPORTED_PKCE_CHALLENGE_METHOD);
+        }
+    }
+
+    public boolean isValidPKCE(String codeVerifier) {
+        if(StringUtils.isEmpty(codeVerifier)) {
+            log.error("Null or empty code_verifier found in the request");
+            throw new EsignetException(ErrorConstants.INVALID_PKCE_CODE_VERFIER);
+        }
+
+        switch (this.codeChallengeMethod) {
+            case S256 :
+                byte[] verifierBytes = codeVerifier.getBytes(Charset.forName("US-ASCII"));
+                String codeVerifierHash = IdentityProviderUtil.generateB64EncodedHash(IdentityProviderUtil.ALGO_SHA_256, verifierBytes);
+                return codeVerifierHash.equals(this.codeChallenge);
+        }
+        throw new EsignetException(ErrorConstants.INVALID_PKCE_CODE_VERFIER);
```

Comment:
As we have ExceptionHandler controller advice caller don't have to handle it explicitly, it will be taken care by exception handler advice.

Reply:
My question is more about the design of this method, I see there 2 possible output when for the same result of validation failed. One is boolean being false and the another is exception. Can't we avoid the boolean and return void and always expect exception in case failure ?

Reply:
Done vishwanath